overlord/overlord.go

```diff
@@ -209,7 +210,46 @@ func (o *Overlord) addManager(mgr StateManager) {
 	o.stateEng.AddManager(mgr)
 }
 
-func loadState(backend state.Backend, restartHandler restart.Handler) (*state.State, error) {
+func initStateFileLock() (*osutil.FileLock, error) {
+	lockFilePath := dirs.SnapStateLockFile
+	if err := os.MkdirAll(filepath.Dir(lockFilePath), 0755); err != nil {
+		return nil, err
+	}
+
+	return osutil.NewFileLockWithMode(lockFilePath, 0644)
+}
+
+func (o *Overlord) loadState(backend state.Backend, restartHandler restart.Handler) (*state.State, error) {
+	flock, err := initStateFileLock()
+	if err != nil {
+		return nil, fmt.Errorf("fatal: error opening lock file: %v", err)
+	}
+	o.flock = flock
+
+	logger.Noticef("Acquiring state lock file")
+	// Setup a timer to print a message at regular intervals. This will help
+	// debugging contentious situations.
+	gotLockCh := make(chan struct{})
+	go func() {
+		ticker := time.NewTicker(time.Minute)
+		defer ticker.Stop()
+		start := time.Now()
+		select {
+		case <-ticker.C:
+			logger.Noticef("Still waiting for state lock file after %s", time.Since(start))
+		case <-gotLockCh:
+			return
+		}
+	}()
+
+	// This will cause the process to block, if the lock is currently held
+	if err := o.flock.Lock(); err != nil {
+		logger.Noticef("Failed to lock state file")
+		return nil, fmt.Errorf("fatal: could not lock state file: %v", err)
+	}
+	gotLockCh <- struct{}{}
```

Comment:
you don't need to send anything, just `close(gotLockCh)` to have the desired effect of waking up the goroutine.